projector-server/src/main/kotlin/org/jetbrains/projector/server/ProjectorServer.kt

```diff
@@ -90,35 +87,70 @@ import kotlin.properties.Delegates
-    builder.onStart = {
-      updateThread = thread(isDaemon = true) {
-        // TODO: remove this thread: encapsulate the logic in an extracted class and maybe even don't use threads but coroutines' channels
-        logger.debug { "Daemon thread starts" }
-        while (!Thread.currentThread().isInterrupted) {
-          try {
-            val dataToSend = createDataToSend()  // creating data even if there are no clients to avoid memory leaks
+  private lateinit var updateThread: Thread
 
-            sendPictures(dataToSend)
+  private val caretInfoQueue = ConcurrentLinkedQueue<ServerCaretInfoChangedEvent.CaretInfoChange>()
 
-            Thread.sleep(10)
-          }
-          catch (ex: InterruptedException) {
-            Thread.currentThread().interrupt()
-          }
-          catch (t: Throwable) {
-            logger.error(t) { "Unhandled in daemon thread has happened" }
-          }
+  private val caretInfoUpdater = CaretInfoUpdater { caretInfo ->
+    caretInfoQueue.add(caretInfo)
+  }
+
+  private val markdownQueue = ConcurrentLinkedQueue<ServerMarkdownEvent>()
+
+  private var windowColorsEvent: ServerWindowColorsEvent? = null
+
+  private val ideaColors = IdeColors { colors ->
+    windowColorsEvent = ServerWindowColorsEvent(colors)
+  }
+
+  init {
+    PanelUpdater.showCallback = { id, show ->
+      markdownQueue.add(ServerMarkdownEvent.ServerMarkdownShowEvent(id, show))
+    }
+    PanelUpdater.resizeCallback = { id, size ->
+      markdownQueue.add(ServerMarkdownEvent.ServerMarkdownResizeEvent(id, size.toCommonIntSize()))
+    }
+    PanelUpdater.moveCallback = { id, point ->
+      markdownQueue.add(ServerMarkdownEvent.ServerMarkdownMoveEvent(id, point.shift(PGraphicsEnvironment.defaultDevice.clientShift)))
+    }
+    PanelUpdater.disposeCallback = { id ->
+      markdownQueue.add(ServerMarkdownEvent.ServerMarkdownDisposeEvent(id))
+    }
+    PanelUpdater.placeToWindowCallback = { id, rootComponent ->
+      rootComponent?.let {
+        val peer = AWTAccessor.getComponentAccessor().getPeer<ComponentPeer>(it)
+
+        if (peer !is PComponentPeer) {
+          return@let
         }
-        logger.debug { "Daemon thread finishes" }
+
+        markdownQueue.add(ServerMarkdownEvent.ServerMarkdownPlaceToWindowEvent(id, peer.pWindow.id))
       }
+    }
+    PanelUpdater.setHtmlCallback = { id, html ->
+      markdownQueue.add(ServerMarkdownEvent.ServerMarkdownSetHtmlEvent(id, html))
+    }
+    PanelUpdater.setCssCallback = { id, css ->
+      markdownQueue.add(ServerMarkdownEvent.ServerMarkdownSetCssEvent(id, css))
+    }
+    PanelUpdater.scrollCallback = { id, offset ->
+      markdownQueue.add(ServerMarkdownEvent.ServerMarkdownScrollEvent(id, offset))
+    }
+    PDesktopPeer.browseUriCallback = { link ->
+      markdownQueue.add(ServerMarkdownEvent.ServerMarkdownBrowseUriEvent(link))
+    }
+  }
+
+  private fun initTransport(): HttpWsTransport {
+    val builder = createTransportBuilder()
+    builder.onStart = {
```

Comment:
This callback is empty now? Do you think we should remove this field?

Reply:
Good idea. onStart removed from HttpWsTransport && TransportBuilder interfaces in this: https://github.com/JetBrains/projector-client/pull/27 PR